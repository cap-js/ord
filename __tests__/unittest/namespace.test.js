const cds = require("@sap/cds");
const { AUTHENTICATION_TYPE } = require("../../lib/constants");

jest.spyOn(cds, "context", "get").mockReturnValue({
    authConfig: {
        types: [AUTHENTICATION_TYPE.Open],
    },
});
const { createAPIResourceTemplate, createEventResourceTemplate } = require("../../lib/templates");

describe("namespace local and global", () => {
    it("should strip application namespace from local namespace", () => {
        const appConfig = {
            ordNamespace: "customer.testNamespace",
            appName: "testAppName",
            lastUpdate: "2022-12-19T15:47:04+00:00",
        };
        const serviceName = "MyService";
        const testNamespace = "customer.testNamespace.nested.";
        const model = cds.linked(`
                namespace customer.testNamespace.nested;

                service MyService {
                    entity Books {
                        key ID: UUID;
                        title: String;
                    }
                };
                annotate MyService with @ORD.Extensions : {
                    visibility : 'public'
                };
            `);
        const packageIds = ["sap.test.cdsrc.sample:package:test-event:v1", "sap.test.cdsrc.sample:package:test-api:v1"];
        const srvDefinition = model.definitions[testNamespace + serviceName];
        expect(createAPIResourceTemplate(serviceName, srvDefinition, appConfig, packageIds)).toMatchSnapshot();
        expect(createEventResourceTemplate(serviceName, srvDefinition, appConfig, packageIds)).toMatchSnapshot();
    });

    it("should strip application namespace if its the same as local namespace", () => {
        const appConfig = {
            ordNamespace: "customer.testNamespace",
            appName: "testAppName",
            lastUpdate: "2022-12-19T15:47:04+00:00",
        };
        const serviceName = "MyService";
        const testNamespace = "customer.testNamespace.";
        const model = cds.linked(`
                namespace customer.testNamespace;

                service MyService {
                    entity Books {
                        key ID: UUID;
                        title: String;
                    }
                };
                annotate MyService with @ORD.Extensions : {
                    visibility : 'public'
                };
            `);
        const packageIds = ["sap.test.cdsrc.sample:package:test-event:v1", "sap.test.cdsrc.sample:package:test-api:v1"];
        const srvDefinition = model.definitions[testNamespace + serviceName];
        expect(createAPIResourceTemplate(serviceName, srvDefinition, appConfig, packageIds)).toMatchSnapshot();
        expect(createEventResourceTemplate(serviceName, srvDefinition, appConfig, packageIds)).toMatchSnapshot();
    });

    it("should not strip a different local namespace", () => {
        const appConfig = {
            ordNamespace: "customer.testNamespace",
            appName: "testAppName",
            lastUpdate: "2022-12-19T15:47:04+00:00",
        };
        const serviceName = "MyService";
        const testNamespace = "other.namespace.";
        const model = cds.linked(`
                namespace other.namespace;

                service MyService {
                    entity Books {
                        key ID: UUID;
                        title: String;
                    }
                };
                annotate MyService with @ORD.Extensions : {
                    visibility : 'public'
                };
            `);
        const packageIds = ["sap.test.cdsrc.sample:package:test-event:v1", "sap.test.cdsrc.sample:package:test-api:v1"];
        const srvDefinition = model.definitions[testNamespace + serviceName];
        expect(createAPIResourceTemplate(serviceName, srvDefinition, appConfig, packageIds)).toMatchSnapshot();
        expect(createEventResourceTemplate(serviceName, srvDefinition, appConfig, packageIds)).toMatchSnapshot();
    });
});
