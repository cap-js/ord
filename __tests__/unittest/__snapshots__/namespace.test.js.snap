// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`namespace local and global should not strip a different local namespace 1`] = `
[
  {
    "apiProtocol": "odata-v4",
    "description": "Description for MyService",
    "entryPoints": [
      "/odata/v4/my",
    ],
    "extensible": {
      "supported": "no",
    },
    "lastUpdate": "2022-12-19T15:47:04+00:00",
    "ordId": "customer.testNamespace:apiResource:other.namespace.MyService:v1",
    "partOfGroups": [
      "sap.cds:service:customer.testNamespace:other.namespace.MyService",
    ],
    "partOfPackage": "sap.test.cdsrc.sample:package:test-api:v1",
    "releaseStatus": "active",
    "resourceDefinitions": [
      {
        "accessStrategies": [
          {
            "type": "open",
          },
        ],
        "mediaType": "application/json",
        "type": "openapi-v3",
        "url": "/ord/v1/customer.testNamespace:apiResource:other.namespace.MyService:v1/MyService.oas3.json",
      },
      {
        "accessStrategies": [
          {
            "type": "open",
          },
        ],
        "mediaType": "application/xml",
        "type": "edmx",
        "url": "/ord/v1/customer.testNamespace:apiResource:other.namespace.MyService:v1/MyService.edmx",
      },
    ],
    "shortDescription": "Short description of MyService",
    "title": "MyService",
    "version": "1.0.0",
    "visibility": "public",
  },
]
`;

exports[`namespace local and global should not strip a different local namespace 2`] = `
[
  {
    "description": "CAP Event resource describing events / messages.",
    "extensible": {
      "supported": "no",
    },
    "lastUpdate": "2022-12-19T15:47:04+00:00",
    "ordId": "customer.testNamespace:eventResource:other.namespace.MyService:v1",
    "partOfGroups": [
      "sap.cds:service:customer.testNamespace:other.namespace.MyService",
    ],
    "partOfPackage": "sap.test.cdsrc.sample:package:test-event:v1",
    "releaseStatus": "active",
    "resourceDefinitions": [
      {
        "accessStrategies": [
          {
            "type": "open",
          },
        ],
        "mediaType": "application/json",
        "type": "asyncapi-v2",
        "url": "/ord/v1/customer.testNamespace:eventResource:other.namespace.MyService:v1/MyService.asyncapi2.json",
      },
    ],
    "shortDescription": "MyService event resource",
    "title": "ODM testAppName Events",
    "version": "1.0.0",
    "visibility": "public",
  },
]
`;

exports[`namespace local and global should strip application namespace from local namespace 1`] = `
[
  {
    "apiProtocol": "odata-v4",
    "description": "Description for MyService",
    "entryPoints": [
      "/odata/v4/my",
    ],
    "extensible": {
      "supported": "no",
    },
    "lastUpdate": "2022-12-19T15:47:04+00:00",
    "ordId": "customer.testNamespace:apiResource:nested.MyService:v1",
    "partOfGroups": [
      "sap.cds:service:customer.testNamespace:nested.MyService",
    ],
    "partOfPackage": "sap.test.cdsrc.sample:package:test-api:v1",
    "releaseStatus": "active",
    "resourceDefinitions": [
      {
        "accessStrategies": [
          {
            "type": "open",
          },
        ],
        "mediaType": "application/json",
        "type": "openapi-v3",
        "url": "/ord/v1/customer.testNamespace:apiResource:nested.MyService:v1/MyService.oas3.json",
      },
      {
        "accessStrategies": [
          {
            "type": "open",
          },
        ],
        "mediaType": "application/xml",
        "type": "edmx",
        "url": "/ord/v1/customer.testNamespace:apiResource:nested.MyService:v1/MyService.edmx",
      },
    ],
    "shortDescription": "Short description of MyService",
    "title": "MyService",
    "version": "1.0.0",
    "visibility": "public",
  },
]
`;

exports[`namespace local and global should strip application namespace from local namespace 2`] = `
[
  {
    "description": "CAP Event resource describing events / messages.",
    "extensible": {
      "supported": "no",
    },
    "lastUpdate": "2022-12-19T15:47:04+00:00",
    "ordId": "customer.testNamespace:eventResource:nested.MyService:v1",
    "partOfGroups": [
      "sap.cds:service:customer.testNamespace:nested.MyService",
    ],
    "partOfPackage": "sap.test.cdsrc.sample:package:test-event:v1",
    "releaseStatus": "active",
    "resourceDefinitions": [
      {
        "accessStrategies": [
          {
            "type": "open",
          },
        ],
        "mediaType": "application/json",
        "type": "asyncapi-v2",
        "url": "/ord/v1/customer.testNamespace:eventResource:nested.MyService:v1/MyService.asyncapi2.json",
      },
    ],
    "shortDescription": "MyService event resource",
    "title": "ODM testAppName Events",
    "version": "1.0.0",
    "visibility": "public",
  },
]
`;

exports[`namespace local and global should strip application namespace if its the same as local namespace 1`] = `
[
  {
    "apiProtocol": "odata-v4",
    "description": "Description for MyService",
    "entryPoints": [
      "/odata/v4/my",
    ],
    "extensible": {
      "supported": "no",
    },
    "lastUpdate": "2022-12-19T15:47:04+00:00",
    "ordId": "customer.testNamespace:apiResource:MyService:v1",
    "partOfGroups": [
      "sap.cds:service:customer.testNamespace:MyService",
    ],
    "partOfPackage": "sap.test.cdsrc.sample:package:test-api:v1",
    "releaseStatus": "active",
    "resourceDefinitions": [
      {
        "accessStrategies": [
          {
            "type": "open",
          },
        ],
        "mediaType": "application/json",
        "type": "openapi-v3",
        "url": "/ord/v1/customer.testNamespace:apiResource:MyService:v1/MyService.oas3.json",
      },
      {
        "accessStrategies": [
          {
            "type": "open",
          },
        ],
        "mediaType": "application/xml",
        "type": "edmx",
        "url": "/ord/v1/customer.testNamespace:apiResource:MyService:v1/MyService.edmx",
      },
    ],
    "shortDescription": "Short description of MyService",
    "title": "MyService",
    "version": "1.0.0",
    "visibility": "public",
  },
]
`;

exports[`namespace local and global should strip application namespace if its the same as local namespace 2`] = `
[
  {
    "description": "CAP Event resource describing events / messages.",
    "extensible": {
      "supported": "no",
    },
    "lastUpdate": "2022-12-19T15:47:04+00:00",
    "ordId": "customer.testNamespace:eventResource:MyService:v1",
    "partOfGroups": [
      "sap.cds:service:customer.testNamespace:MyService",
    ],
    "partOfPackage": "sap.test.cdsrc.sample:package:test-event:v1",
    "releaseStatus": "active",
    "resourceDefinitions": [
      {
        "accessStrategies": [
          {
            "type": "open",
          },
        ],
        "mediaType": "application/json",
        "type": "asyncapi-v2",
        "url": "/ord/v1/customer.testNamespace:eventResource:MyService:v1/MyService.asyncapi2.json",
      },
    ],
    "shortDescription": "MyService event resource",
    "title": "ODM testAppName Events",
    "version": "1.0.0",
    "visibility": "public",
  },
]
`;
